### PR TITLE
Swift: Query for SQL injection

### PR DESCRIPTION
**Pull request by geoffw0**
Query for SQL injection in Swift.

~Draft: currently there are just test cases, covering the [C API](https://www.sqlite.org/c3ref/funclist.html) that can be [used in Swift](https://developer.apple.com/documentation/swift/using-imported-c-functions-in-swift), and the [SQLite.swift](https://github.com/stephencelis/SQLite.swift) wrapper that I believe is commonly used as well.  I'd like someone to review these test cases for realism / misconceptions / any important details I've missed before going too much further.~

@atorralba would you mind taking a look?

### CODE REVIEW

File: swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift
Comment (atorralba) [inline swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift:155]: This is safe indeed. That's the proper way of building SQL statements with untrusted data: using prepared statements and binding the untrusted data to the statement with the appropriate type.

File: swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift
Comment (atorralba) [inline swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift:98]: I'm not sure how `sanitize` should work, though. Is it a test-only function to test sanitizers in general, or do we expect the query to recognize any function called `sanitize` or similar as a valid sanitizer?

**Review by atorralba — COMMENTED** [review]
The tests make sense to me. `sqlite3_exec` and `sqlite3_prepare*` seem to be the appropriate sinks, because they handle raw SQL statements. `sqlite3_bind*` handle the parameters of prepared statements, which is safe to do even with untrusted data.

This should cover the raw `sqlite3` library. Are we planning on covering wrapper libraries on top of it? e.g. https://github.com/stephencelis/SQLite.swift or https://github.com/groue/GRDB.swift.

**Comment by geoffw0** [comment]
> Are we planning on covering wrapper libraries on top of it? e.g. https://github.com/stephencelis/SQLite.swift

The tests in `SQLite.swift` should cover this library - assuming I've correctly identified the right / most important interfaces.

> or https://github.com/groue/GRDB.swift.

Yeah it would be good to cover this library as well.  However I'm a bit behind on this task as it is, so I'll create a follow-up ticket rather than addressing this now.

File: swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift
Comment (geoffw0) [reply swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift:155]: Thanks, good to confirm this is the right way.

File: swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift
Comment (geoffw0) [reply swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift:98]: The query is likely to recognize functions using a heuristic such as the function name, so matching functions called `sanitize`, `filter` etc (`escape`, perhaps).  The tricky part is that's likely to include some bad sanitizers (like just adding quotes to the ends of the string, or escaping some but not all important characters) that we would rather flag somehow.

Comment (geoffw0) [reply swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift:98]: I see that in Java we have a different approach - `PrimitiveType`, `BoxedType`, `NumberType` are sanitizers.  I assume this is because they can't possibly contain or expand to free text as would be required for an injection attack.

Comment (geoffw0) [reply swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift:98]: I'm unclear whether it's *possible* to write a safe sanitizer for SQL, or whether in fact the prepared statement approach is the only 'right' way?

Comment (atorralba) [reply swift/ql/test/query-tests/Security/CWE-089/sqlite3_c_api.swift:98]: You _technically_ could, but it's not recommended because it's easy to do it wrong. In general, the suggested approach always is using prepared statements (or wrapper libraries that do not handle raw SQL) when possible, or otherwise using a well-established security library to do the sanitization/validation for you.

File: swift/ql/test/query-tests/Security/CWE-089/SQLite.swift
Comment (atorralba) [inline swift/ql/test/query-tests/Security/CWE-089/SQLite.swift:75]: Same reason as in `sqlite3_c_api.swift`, this is safe.

File: swift/ql/test/query-tests/Security/CWE-089/SQLite.swift
Comment (atorralba) [inline swift/ql/test/query-tests/Security/CWE-089/SQLite.swift:78]: This should include one of `unsafeQuery*` being used as first parameter of the multi-parameter versions of `prepare()`, which would be `BAD`.

I guess you are already aware, but `Connection.(run|bind|scalar)`'s first argument is a sink, whereas `Statement.(bind|run|scalar)`'s arguments are all safe. The string parameter of `Statement`'s constructor is also a sink.

**Review by atorralba — COMMENTED** [review]
> The tests in SQLite.swift should cover this library - assuming I've correctly identified the right / most important interfaces.

D'oh, true, sorry.

**Comment by geoffw0** [comment]
I've removed draft status from this PR, it's ready to review.

I'd like to acknowledge a few issues though.  IMO none of these prevent us from merging what we have:
 - we're getting no results on the C API test due to an issue with taint flow, possibly itself caused by an extractor issue.  I'm discussing this with other members of the team and I'm not sure if the fix will be something simple I can do here, or something that will need to be done as follow-up.
 - there are no sanitizers for remote data in numerical form yet (considered to be safe).  This is because in my test case taint doesn't flow into `remoteNumber` anyway, so there's no way to test such a sanitizer (and no strict need for it yet).
 - we may also want sanitizers along the lines of Ruby's `StringConstCompareBarrier` and `StringConstArrayInclusionCallBarrier`.  I might write a follow-up issue for this but I'm inclined not to worry too much until we see patterns like those in real Swift code (they are talked about in https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html).

**Comment by geoffw0** [comment]
I've just merged main into this PR - with the extractor fix https://github.com/github/codeql/pull/10854 that gives us the C API results for this query.  We still don't get results through the UTF-16 interface, but that's much less concerning than missing all C API results.

I'd be happy to see this reviewed and merged, and deal with the remaining known issues as follow-up.

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.ql
Comment (jketema) [inline swift/ql/src/queries/Security/CWE-089/SqlInjection.ql:62]: Just for my understanding: `init` is the name assigned to the constructor by the extractor?

File: swift/ql/test/query-tests/Security/CWE-089/SQLite.swift
Comment (jketema) [inline swift/ql/test/query-tests/Security/CWE-089/SQLite.swift:130]: These I do not quite understand, even after looking at the SQLite.swift docs. Shouldn't these be tagged with `// GOOD`, or am I missing something?

**Review by jketema — APPROVED** [review]
Two small questions. Otherwise this looks ok to me.

**Review by jketema — COMMENTED** [review]
I'm happy with the current version, so just waiting for the doc review.

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp:8]:
```suggestion
If a database query (such as a SQL query) is built from user-provided data without sufficient sanitization, a user may be able to run malicious database queries.
```

[MS Style guide treats as an acronym](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/s/sql-sql-server) ("a SQL...")

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp:21]:
```suggestion
<p>In the following example, a SQL query is prepared using string interpolation to directly include a user-controlled value <code>userControlledString</code> in the query. An attacker could craft the part they control to change the overall meaning of the SQL query.
```

(as above)

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp:26]: Is there a different way of expressing the "break out" part? Unless this is a technical term related to SQL then it'd be preferable to avoid metaphors and expressions (basically anything in inverted commas) so as to help comprehension, particularly for people for whom English is a second language 🙏

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp:21]: I hope this isn't adding too much overhead but could you kindly explain what's meant by "the part they control" (the attacker)? The reason I ask is because this paragraph feels like the central part of the whole piece and some feedback I have might or might not be relevant.

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp:15]: In the context, is "user-provided data without sufficient sanitization" equivalent to "building queries by string concatenation"?

**Review by subatoi — COMMENTED** [review]
Hi there @geoffw0, I'm the first responder for the Docs team this week so I've taken a look at this and left some small comments and asked a couple of minor questions. I hope this isn't adding too much overhead 🙏 

I'll just tag my colleague @felicitymay as well just in case (might not be necessary but I'm still quite new to reviewing CodeQL! 😁 )

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (geoffw0) [reply swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp]: Ah, I've seen that elsewhere and thought it was a mistake!  I've accepted your changes.

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (geoffw0) [reply swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp:15]: Yes this is the same idea, though they're not quite formally equivalent.  Concatenation will be by far the most common case.  I'll think about whether we can be more consistent describing this.

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (geoffw0) [reply swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp]: It means the part of the concatenated string that comes from the user.  I agree it may be worth moving this bit to sooner in the document.

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (geoffw0) [reply swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp]: Yeah, I can see that this might be confusing for someone who doesn't speak English as a first language.  I've removed the metaphor.

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (geoffw0) [reply swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp]: I've added something very similar to this point to the opening paragraph of the doc.  There's a bit of repetition but that may not be a bad thing.  I've also rephrased "the part they control" to be a bit more explicit.

**Comment by geoffw0** [comment]
Thanks for the review @subatoi .  I believe I've addressed all your points, but please don't hesitate to say if you think it needs more work.

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp:15]:
```suggestion
Most database connector libraries offer a way to safely embed untrusted data into a query using query parameters or prepared statements. You should use these features to build queries, rather than string concatenation or similar methods without sufficient sanitization.
```

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp:21]:
```suggestion
<p>In the following example, a SQL query is prepared using string interpolation to directly include a user-controlled value <code>userControlledString</code> in the query. An attacker could craft <code>userControlledString</code> to change the overall meaning of the SQL query.
```

**Review by subatoi — APPROVED** [review]
Just added a couple of minor last points to make the language a bit simpler. There's nothing wrong with the existing language; these suggestions are just minor and style-related. Otherwise approving and LGTM. Thank you!

File: swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp
Comment (geoffw0) [reply swift/ql/src/queries/Security/CWE-089/SqlInjection.qhelp:21]: Sometimes no words are the best words. :)

**Comment by geoffw0** [comment]
Thanks - those suggestions definitely help make things read a bit more easily.  Please approve the PR if you're happy with the docs now (the code has already been reviewed and approved).

**Review by subatoi — APPROVED** [review]
Thanks @geoffw0, apologies, didn't realise my review would be dismissed. LGTM!

**Comment by geoffw0** [comment]
Thanks again for the review, suggestions and approval.  On to the next query!

**Pull request merged**
